src/DataStreams/ColumnGathererStream.h

```diff
@@ -96,11 +101,13 @@ class ColumnGathererStream : public IBlockInputStream
 
     std::vector<Source> sources;
     ReadBuffer & row_sources_buf;
+    bool enable_low_cardinality_merge_new_algo;
 
     size_t block_preferred_size;
 
     Source * source_to_fully_copy = nullptr;
     Block output_block;
+    MutableColumnPtr cardinalityDict = nullptr;
```

Comment:
The idea of optimisation is good and clear: let's store a common dictionary, and let's transform all source columns the way it use this common dict; so that now we have same indexes and may update only them.

Here is my suggestion how we can probably simplify implementation (so we won't need to add anything new to ColumnUnique and ColumnLowCardinality):
* Common dictionary ('cardinalityDict' here) may be stored directly in `ColumnUnique`. In `gatherLowCardinality` we can build indexes separately, and then return `ColumnLowCardinality::create(cardinalityDict, indexes)`.
* For each source column, let's support it's indexes in a common dict. At a time we read a column, those indexes will be a result of `cardinalityDict->uniqueInsertRangeFrom(column.getDictionary().getNestedColumn(), 0, size)->index(column.getIndexes());` For simplicity, we can additionaly cast this result to UInt64.
* We can reuse existing implementation of `gather` just for indexes (at least, most of the parts). Or, maybe, add a couple of `if (low_cardinality)` to existing one.

Reply:
Great! seems more clear. But I cannot make sure that this simplify can have a better performance. I will try that!

Reply:
@KochetovNicolai Hi, if I understand you correctly, this simplify implementation may not work correctly.
Here **cardinalityDict->uniqueInsertRangeFrom(column.getDictionary().getNestedColumn(), 0, size)->index(column.getIndexes())**, the cardinalityDict is shared, and the index returned from uniqueInsertRangeFrom not aligned with the column original index column.getIndexes(), So before you need do the remapping **auto idx_map = mapUniqueIndex(*sub_idx)** and change the  sub_idx align with the unique insert returned index. Am I right?